lib/storage/data/file/store.js

```diff
@@ -0,0 +1,274 @@
+'use strict'; // eslint-disable-line
+
+const fs = require('fs');
+const crypto = require('crypto');
+const async = require('async');
+const assert = require('assert');
+
+const Logger = require('werelogs').Logger;
+const errors = require('../../../errors');
+const stringHash = require('../../../stringHash');
+const jsutil = require('../../../jsutil');
+const storageUtils = require('../../utils');
+
+const FOLDER_HASH = 3511;
+
+
+/**
+ * @class
+ * @classdesc File-based object blob store
+ *
+ * Each object/part becomes a file and the files are stored in a
+ * directory hash structure under the configured dataPath.
+ */
+class DataFileStore {
+
+    /**
+     * @constructor
+     * @param {Object} dataConfig - configuration of the file backend
+     * @param {String} dataConfig.dataPath - absolute path where to
+     *   store the files
+     * @param {Boolean} [dataConfig.noSync=false] - If true, disable
+     *   sync calls that ensure files and directories are fully
+     *   written on the physical drive before returning an
+     *   answer. Used to speed up unit tests, may have other uses.
+     * @param {Object} logConfig - logging configuration
+     */
+    constructor(dataConfig, logConfig) {
+        this.logger = new Logger('DataFileStore', logConfig);
+        this.dataPath = dataConfig.dataPath;
+        this.noSync = dataConfig.noSync || false;
+    }
+
+    /**
+     * Setup the storage backend before starting to read or write
+     * files in it.
+     *
+     * The function ensures that dataPath is accessible and
+     * pre-creates the directory hashes under dataPath.
+     *
+     * @param {function} callback - called when done with no argument
+     * @return {undefined}
+     */
+    setup(callback) {
+        fs.accessSync(this.dataPath, fs.F_OK | fs.R_OK | fs.W_OK);
+
+        // Create FOLDER_HASH subdirectories
+        const subDirs = Array.from({ length: FOLDER_HASH },
+                                   (v, k) => (k).toString());
+        this.logger.info(`pre-creating ${subDirs.length} subdirs...`);
+        if (!this.noSync) {
+            storageUtils.setDirSyncFlag(this.dataPath, this.logger);
+        }
+        async.eachSeries(subDirs, (subDirName, next) => {
+            fs.mkdir(`${this.dataPath}/${subDirName}`, err => {
+                // If already exists, move on
+                if (err && err.code !== 'EEXIST') {
+                    return next(err);
+                }
+                return next();
+            });
+        },
+        err => {
+            assert.strictEqual(err, null, `Error creating data files ${err}`);
```

Comment:
It's not ideal to throw in async functions. I recommend sending the error to the caller and taking care of it.